### PR TITLE
feat(system_error_monitor): output emergency reason for debugging

### PR DESCRIPTION
**Pull request by takayuki5168**
## Related Issue(required)

Resolves https://github.com/autowarefoundation/autoware.universe/issues/508

## Description(required)

output latent fault (as rclcpp warn) and single point fault (as rclcpp error) in system_error_monitor

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [commit-guidelines][commit-guidelines]
- [x] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [x] Commits are properly organized and messages are according to the guideline
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by takayuki5168** [comment]
- Just launching autoware with simple psim, no debug message is output every 5s cycle.
- After positioning the ego vehicle, debug message is output every 5s cycle as follows
image
- After putting the goal pose, no debug message is output every cycle.

- When putting the ego out of lane, single point fault of lane_departure_checker occurs, and debug message is output every 5s cycle as follows
image

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (kenji-miyake) [inline system/system_error_monitor/src/system_error_monitor_core.cpp:122]: Since there are cascading failures, I guess the root cause might not appear in the log with `THROTTLE`.

Is it okay?
Also, could you test this behavior? You can use `dummy_diag_publisher`.

Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp:122]: @kenji-miyake 
I've tested some cases, but I'll test more with `dummy_diag_publisher`.
https://github.com/autowarefoundation/autoware.universe/pull/510#issuecomment-1065316513

I wanted to show the root(?) cause, but lane_departure_checker just outputs `Error`. `vehicle is out of the lane` is better, which I guess is the root cause.
 I confirmed that without THROTTLE, I can output `vehilce is out of the lane`. I really appreciate your comment.

I'm not familiar with cascading failure structure, but is there any way to output just the root cause with THROTTLE?

Comment (kenji-miyake) [reply system/system_error_monitor/src/system_error_monitor_core.cpp:122]: Mm... I meant you have to trigger two errors at the same time.
But in this case, a different diagnostic (the parent of your target diagnostic) was coincidentally displayed, it might be okay.

> I'm not familiar with cascading failure structure, but is there any way to output just the root cause with THROTTLE?

I remember it was difficult.
We have to name each `THROTTLE` a different name.

Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp:122]: I changed the name of the logger of RCLCPP_ERROR_THROTTLE for the parent and cause reason, but only the parent reason is output. If there seems no good way to output the root reason, I give up for now.

> Mm... I meant you have to trigger two errors at the same time.
Ah, you mean I should try two different errors at the same time. OK, I will.

Comment (kenji-miyake) [reply system/system_error_monitor/src/system_error_monitor_core.cpp:122]: I remember, with the current implementation of `THROTTLE`, you have to decide the names during the build phase, not the runtime phase.

Comment (kenji-miyake) [reply system/system_error_monitor/src/system_error_monitor_core.cpp:122]: Please confirm the source of the macros.

Comment (kenji-miyake) [reply system/system_error_monitor/src/system_error_monitor_core.cpp:122]: Or perhaps I've implemented a hacky way in `ad_service_state_monitor`? :thinking:

Comment (kenji-miyake) [reply system/system_error_monitor/src/system_error_monitor_core.cpp:122]: I've found out it.
https://github.com/tier4/AutowareArchitectureProposal.iv/blob/30f2ca0a82e836e2d787264326af6cf6c832d4b2/system/autoware_state_monitor/include/autoware_state_monitor/rosconsole_wrapper.h#L28

Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp:122]: I don't understand the macro of THROTTLE

Thank you. Then I will port logThrottleNamed in another PR.

Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: @kenji-miyake 
Finally I could implement throttle log output considering the logger name. For now I put them in the same file as follows.

https://github.com/autowarefoundation/autoware.universe/blob/eeb8406ad10c170943a919c35bdb930742eb76d2/system/system_error_monitor/src/system_error_monitor_core.cpp#L29-L48

**Comment by takayuki5168** [comment]
- a bit long error output every 5 sec, after positioning ego pose
image

- put ego pose out of lane, and error output every 5 sec for lane_departure_checker
image

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: It works well (I put the screenshot in the bottom). I don't use dummy_diag_publisher though.

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (kenji-miyake) [inline system/system_error_monitor/src/system_error_monitor_core.cpp:44]: It's better to replace this with `RCUTILS_LOG_COND_NAMED` or `template <int log_level>`.

Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: `RCUTILS_LOG_COND_NAMED` requires `RCUTILS_LOG_CONDITION_ONCE_BEFORE`, `RCUTILS_LOG_CONDITION_ONCE_AFTER`, ... which makes the readability less, so I used if else.

I'm not sure about `if constexpr`, but is this code what you intended to do?

Comment (kenji-miyake) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: > RCUTILS_LOG_COND_NAMED requires RCUTILS_LOG_CONDITION_ONCE_BEFORE, RCUTILS_LOG_CONDITION_ONCE_AFTER, ... which makes the readability less

Could you explain this in a bit more detail?

> I'm not sure about if constexpr, but is this code what you intended to do?

Yes, almost perfect.
But it's better to use `enum class` instead of `int`.
Or if there are some macros like `RCLCPP_LOG_LEVEL_DEBUG`, you can use it instead.

Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: Arguments of `RCUTILS_LOG_COND_NAMED` is like this.
image
, and this is an example of macros for before and after.
image

Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: >Or if there are some macros like RCLCPP_LOG_LEVEL_DEBUG, you can use it instead.

There is defined values like this, which I don't think good to use in the application side, so I didn't use them.
image

Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/510/commits/a767e1aa6e03e5eb53b326ca27b6221265855927
DONE

Comment (kenji-miyake) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: > , and this is an example of macros for before and after.

Hmm? Why `RCUTILS_LOG_CONDITION_ONCE_BEFORE` and `RCUTILS_LOG_CONDITION_ONCE_AFTER` are required?

Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: You mean, it's ok just to put empty for BEFORE and AFTER? If so, it's right. Anyway we have to use flag like `RCUTILS_LOG_MIN_SEVERITY_DEBUG`, which I don't think good as I said.

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (kenji-miyake) [inline system/system_error_monitor/src/system_error_monitor_core.cpp:31]: Seeing https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#enum8-specify-enumerator-values-only-when-necessary,

You have to remove `= 0`,

```suggestion
enum class DebugLevel { DEBUG, INFO, WARN, ERROR };
```

or add some comments?

```suggestion
enum class DebugLevel { DEBUG = 0, INFO, WARN, ERROR }; // To ...
```

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (kenji-miyake) [inline system/system_error_monitor/src/system_error_monitor_core.cpp:33]: Redundant.

```suggestion
```

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (kenji-miyake) [inline system/system_error_monitor/src/system_error_monitor_core.cpp:42]: Adding the unit would be more readable.

```suggestion
    if (time_from_last_output.seconds() * 1000.0 < duration_ms) {
```

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (kenji-miyake) [inline system/system_error_monitor/src/system_error_monitor_core.cpp:56]: It seems there is also the macro `RCLCPP_FATAL`.

https://docs.ros2.org/bouncy/api/rclcpp/logging_8hpp.html#a28cf59074387c81ef082f3e774481e95

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (kenji-miyake) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: > Do you mean it's ok just to put empty for BEFORE and AFTER? If so, it's right. 

Yes, I think it's okay.

> Anyway we have to use flag like RCUTILS_LOG_MIN_SEVERITY_DEBUG, which I don't think good as I said.

I think it's acceptable because it's documented.
https://docs.ros2.org/bouncy/api/rclcpp/logging_8hpp.html#a28cf59074387c81ef082f3e774481e95

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/510/commits/142b5a553a992dc53d0db0231db87439c646f0d9
Done

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/510/commits/142b5a553a992dc53d0db0231db87439c646f0d9
I don't understand the second suggestion, but Doen

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: https://github.com/autowarefoundation/autoware.universe/pull/510/commits/142b5a553a992dc53d0db0231db87439c646f0d9
Done

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp:57]: https://github.com/autowarefoundation/autoware.universe/pull/510/commits/142b5a553a992dc53d0db0231db87439c646f0d9
Done

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (kenji-miyake) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: I meant, for example `To be the same as rclcpp's macros`

**Review by kenji-miyake — APPROVED** [review]
If it's re-tested, LGTM.

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: Just got it.

File: system/system_error_monitor/src/system_error_monitor_core.cpp
Comment (takayuki5168) [reply system/system_error_monitor/src/system_error_monitor_core.cpp]: Just go it.

**Comment by takayuki5168** [comment]
Thank you. I retested, and confirmed it works well.

My concern is after positioning ego pose and before positioning goal pose, a bit long error is output every 5 secs.
How do you feel?
image

**Comment by kenji-miyake** [comment]
@takayuki5168 In that case, you can add some states to ignore.

**Comment by kenji-miyake** [comment]
Ideally, we should extend system_error_monitor to understand which diagnostics can be ignored in each state respectively.
But since it takes time we can ignore them globally (apply the same settings for all diagnostics) for now.

**Comment by takayuki5168** [comment]
I think planning_error_monitor does not have to output error for no messages coming, since whether there are required messages or not is seen in another monitoring.

**Comment by takayuki5168** [comment]
https://github.com/autowarefoundation/autoware.universe/pull/515
I created another PR to suppress error of planning_error_monitor.

I'll merge this PR after planning_error_monitor's PR is merged.

**Pull request merged**
